### PR TITLE
Issue #957: Cleanup package-private methods

### PR DESCRIPTION
**Pull request by mkordas**
Part 1

### CODE REVIEW

**Comment by romani** [comment]
pmd violation: https://travis-ci.org/checkstyle/checkstyle/jobs/84835358#L251

**Comment by mkordas** [comment]
@romani - that's a proof that check is valuable - package-private access was not triggering unused method rule. And apparently in GUI we still have some dead code...

**Comment by romani** [comment]
> that's a proof that check is valuable - package-private access was not triggering unused method rule

ok, even without this prove , it was clear that it is useful

> And apparently in GUI we still have some dead code

remove or suppress for now. Suppression could be resolved in second PR.

**Comment by mkordas** [comment]
@romani, I have TeamCity failure. It's strange - looks like https://youtrack.jetbrains.com/issue/IDEA-107730 is fixed and I shouldn't get these warnings...

**Comment by romani** [comment]
just put a suppression and move forward, As issue become resolved Teamcity will notify us that suppression is not required any more..

**Comment by mkordas** [comment]
I see that it should have been fixed in 2013. I'll try to investigate a bit first.

**Comment by romani** [comment]
I am completely disagree on usage of VisibleForTesting annotation. It does not do any good to code, UTs could use reflection and any other hacks to get access to class methods, no need to keep holes in class design for UTests. We should not use this annotation in our code.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/Guard.java [hunk 1]
```diff
@@ -88,7 +90,7 @@
      * @param forImport the package to check.
      * @return a result {@link AccessResult} indicating whether it can be used.
      */
-    AccessResult verifyImport(final String forImport) {
+    @VisibleForTesting AccessResult verifyImport(final String forImport) {
```
Comment (mkordas) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/Guard.java]: Codacy Issue found: [The method verifyImport() has an NPath complexity of 18](https://www.codacy.com/app/checkstyle/checkstyle/file/1099998019/issues/source?bid=2193633&fileBranchId=2892283#l93)

**Comment by mkordas** [comment]
@romani 
OK, I'm going to throw away commit with `@VisibleForTesting`, but I don't see any other clean solution for cases when visibility is slightly increased for better testability purposes:
* reflection - should be generally avoided, not readable
* refactoring - big effort, should be probably handled as separate issue per each case and spreaded across contributors

**Comment by romani** [comment]
> reflection - should be generally avoided, not readable

increase of visibility is a hack. It is much better to make a hack in test that is always close to code. readability could be improved by usage of some util methods and it will not looks unclear, it will be the same as extra annotation in production code. 
UT should do all testing by public methods, if it is not possible - it is perfect sign that design is not ideal and should be reconsidered. But even case is exceptional and all team members agree on it, it is better to damage Test code rather then production code.

> refactoring - big effort, should be probably handled as 

Please add all cases to excludes and let issue be open. We at least need stop appearance of such visibility in new code. Old code should be fixed with attention.

**Comment by mkordas** [comment]
@romani, by that way of thinking we should also review all usages of `public` and replace with reflection all cases where such public method was made public only for test purposes. Public methods are much more harmful than package-private, so do we put our effort in the right place?

**Comment by romani** [comment]
public methods are not harmful if you design application correctly. Public methods it is allowed/official operations on object, if by means of public methods user can do harm - it is sign of non ideal design (know problems caused by setters to all fields, ..... ).

package-private(some time protected) visibility create a hole for UTs and for clients. After few years they could come to us and demand it as feature to keep their functionality to work (reminder: recent case for non-api changes from public to private ). But in secure code they have no ability to do this and they will start conversation about public way to manage Class status (we have issue from eclipse-cs, I had few conversations in mail-list about hacking over Checkstyle t get access to internals). 

UTs should go the same way as our clients will go to our code. If we design new module, we will write UTs in the same time, massive reflection will be noticed on codereview. As you might noticed a lot of reflection was involved to make 100% coverage (in most cases it is for exceptional cases, as I need to emulate bug JDK :(, .... ).

**Comment by romani** [comment]
Please add known/complicated cases to suppression/exclude and enforce a Rule on the rest of the code.

**Comment by romani** [comment]
merged as FF

**Pull request closed without merging**